sdk/instance/backend.go

```diff
@@ -91,18 +97,21 @@ func (s *Backend) Exists(request cosmostypes.Request, hash hash.Hash) (bool, err
 }
 
 // List returns all instances.
-func (s *Backend) List(request cosmostypes.Request) ([]*instance.Instance, error) {
+func (s *Backend) List(request cosmostypes.Request, f *api.ListInstanceRequest_Filter) ([]*instance.Instance, error) {
 	var (
 		instances []*instance.Instance
 		iter      = request.KVStore(s.storeKey).Iterator(nil, nil)
 	)
 
+	// filter results
 	for iter.Valid() {
 		var i *instance.Instance
 		if err := codec.UnmarshalBinaryBare(iter.Value(), &i); err != nil {
 			return nil, err
 		}
-		instances = append(instances, i)
+		if f == nil || (f != nil && (f.ServiceHash.IsZero() || i.ServiceHash.Equal(f.ServiceHash))) {
```

Comment:
`f` will not be nil if we pass the first 'or', also inverted the last condition to be more consistent with the conditions

```suggestion
		if f == nil || f.ServiceHash.IsZero() || f.ServiceHash.Equal(i.ServiceHash) {
```